### DISCUSSION TITLE
[Bug]: Colormap not working

### DISCUSSION THREAD

**Discussion by vonfzm** [spectrochempy/spectrochempy#487] [Q&A]
### What happened?

I read a folder containing 400 csv files by X=scp.NDDataset.read_csv("irdata"). And when I plot it, only one color was used for all the 400 lines. If I change the colormap, the color will change to the color of the bottom of the colormap, but still all same for all lines.

### Environment

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.15 (default, Oct 12 2022, 19:14:55) 
[GCC 7.5.0]
python-bits: 64
OS: Linux
OS-release: 5.10.133+
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')

quadprog: 0.1.11
brukeropusreader: 1.3
quaternion: 2022.4.2
cantera: None
colorama: 0.4.5
dill: 0.3.5.1
ipython: None
jinja2: 2.11.3
matplotlib: 3.5.3
numba: 0.55.0
numpy: 1.21.5
pint: 0.18
requests: 2.23.0
scipy: 1.7.3
tqdm: 4.64.1
traitlets: 5.1.1
traittypes: 0.2.1
xlrd: 1.1.0
pyyaml: None
ipywidgets: 7.7.1
ipympl: 0.9.2
setuptools: 57.4.0
setuptools_scm: 7.0.5
git: 3.1.29
jupyterlab: None
nodejs: None
pytest: 3.6.4
pytest-doctestplus: None
pytest-flake8: None
pytest-mock: None
pyfakefs: None
scikit-image: None
coverage: None
black: None
pre-commit: None
cffconvert: None
mamba: None
jupytext: None
sphinx: 1.8.6
sphinx_rtd_theme: None
autodocsumm: None
sphinx-gallery: None
nbsphinx: None
jupyter_sphinx: None
json5: None
sphinx-copybutton: None
numpydoc: None
pandoc: None
conda-build: None
conda-verify: None
anaconda-client: None
xarray: 0.20.2
scikit-learn: None
dash: None
dash-bootstrap-components: None
dash: None
daq: None
jupyter-dash: None
plotly: 5.5.0
pip: 21.1.3
autodoc_traits: None
dash_defer_js_import: None
dash-ace: None
spectrochempy: 0.4.7

**Comment by atravert** [comment]
Dear Zimou Feng,

Thank you for your message. This is an expected behaviour. Below I explain why and how to solve it. I hope this helps, but let us know if you have further questions. 

## 1. Why a single color is used ?
 
When plotting a 2D NDDataset, the colormap applies to the coordinates of the first (`y`) dimension of the dataset (let's call it `D`). When reading, e.g., omnic or opus spectra, these coordinates  (in `D.y.data`) are set to the acquisisition timestamp (see [here](https://www.spectrochempy.fr/latest/userguide/importexport/importIR.html#Acquisition-dates-and-y-axis) in the user guide). Since the acqusistion time is not indicated in a csv file, all values of `D.y.data` are set to '0' , hence the same color is used in the plot.  

Here a sample code (and its output) to ilustrate this behaviour:

```python
# load and plot a regular spg file:
A_spg = scp.read("irdata/CO@Mo_Al2O3.SPG")
A_spg.plot()
```
![Figure 1](https://user-images.githubusercontent.com/10098314/196938320-af763c13-cb7b-49d9-a0db-0f78fc3e0993.png)

In the plot above, the color map applies to the y coordinates that are non_zero:
```python
A_spg.y.data
```
returns;
```
array([1.477e+09, 1.477e+09, ..., 1.477e+09, 1.477e+09])
```
Now, let's save each spectrum as a csv file, load them and plot the datast:

```python
for i, spec in enumerate(A_spg):
    s = np.stack((spec.x.data, spec.data.squeeze()), axis=1)
    np.savetxt(f"s_{i}.csv", s, delimiter=";")
A_csv = scp.read_csv(".")
A_csv.plot()
```
![Figure 2](https://user-images.githubusercontent.com/10098314/196940624-e75b4e23-b092-4065-b7fe-e66a0d956111.png)

Here the corlors are identical because all `y` coordiantes are zero:
```python
A_csv.y.data
```
returns;
```
array([0, 0, ..., 0, 0])
```
(note also that as all metada have been lost in the conversion to csv, axis titles, units, etc... are set to default values, and the `x` axis is now plotted in the normal (no-reverse order), because spectrochempy does not recognize an IR spectum) 

## 2. How to solve the problem ?

### Solution 1:  Set the `y` coordinate to distinct values: 

```python
A_csv.y = np.arange(len(A_csv.y))
A_csv.plot()
```
![Figure 3](https://user-images.githubusercontent.com/10098314/196943043-64632819-bde2-4dc8-b434-861fe793957c.png)

```python
A_csv.y.data
```
returns;
```
array([       0,        1, ...,       17,       18])
```
Note that in this case, the color ordering is not the same as for the original dataset: this is because when reading a directory, spectrochempy uses the `pathlib` module which does not sort the files by alphabetiocal order. We could chnage this behavious in a future version.

### 2. Soluton 2: Set `colormap=None`  (dev version only)

We have recently added this option (see the discussion #480).

```python
A_csv.plot(cmap=None)
```
![Figure 4](https://user-images.githubusercontent.com/10098314/196947387-f2d42a90-8f5e-4114-bc44-8bbd7492df31.png)

Note that this option is currently only available in the dev version (see [here](mamba install -c spectrocat/label/dev spectrochempy) to see how to install it), and will be available in the 
next release.

**Reply by atravert** [reply]
PS: note also that if your csv files have been extracted from source file(s) containing the metadata (e.g. the files generated by the spectrometer), it can be worth writing a dedicated reader (see [here](https://www.spectrochempy.fr/latest/devguide/codespec_adding_reader.html#adding-a-reader) in the documentation,). We can help for this as well.

**Reply by vonfzm** [reply]
Dear Arnaud TRAVERT,

Thanks so much for replying, and that works perfect for me.

I have some more questions. Now I have an OPUS (0 file), it is a mapping file containing 400 single IR spectras. But if I read_opus I get only one spectra. 

If I export 400 csv files using OPUS and name the as EXTRACT000.0 ... EXTRACT399.0, then put them in one folder, use read_csv("folder"), do some processing, and I want to export the result to one csv file. But write_csv is only implemented for 1D NDDatasets. If I export the result to jdx file and import it to matalab, they will not list by the original order. Could you please teach me a proper way? I cannot find a right export method in the document, and not much results in search engine. Do we have a group or bbs to chat and seek for help?

Many thanks!

**Reply by atravert** [reply]
Hi,

Can you attach this opus file and a zip of the csv files ? I will work on the opus_reader to handle this type of files. Note that this may take some time... 

Regarding the reading/concatenation of your csv files,  I think that you should use standard python libraries (pathlib and pandas should work fine). I can have a look at this too.
Best

**Reply by vonfzm** [reply]
Hi, 

Sorry for the late replying. I uploaded them the opus file and the zip of the csv files to google drive.
https://drive.google.com/file/d/19_wvyR4IAg5HSdTKG82kobVaFLK-afa2/view?usp=sharing, https://drive.google.com/file/d/1giJxPw8ddszW21XxbTHm1IVWRT51B3Fi/view?usp=sharing

If the links doesn't work for you please tell me.

Thanks a lot!

**Reply by vonfzm** [reply]
Hi, 

Sorry to bother you again. I am having trouble installing scp now, neither on my Mac nor on Google Colab. 

On my M1 Pro Macbook Pro (MacOS 13), I keep getting ERROR: Failed building wheel for numba & ERROR: Failed building wheel for quadprog.

On Google Colab, I was using it yesterday, but today the session will crash at the very beginning section (i.e. !python -m pip install spectrochempy &>> output.log exit(0))    (for both stable and development versions).

If I use jupyterlab, I can install it successfully, but when I run import spectrochempy as scp, I will get "ModuleNotFoundError: No module named 'quaternion'" even if I have already installed quaternion.

Best

**Comment by atravert** [comment]
Hi,
- on google colab:! it should work fine (it takes ~ 1min15 secs). Note that the message below is expected because we have to `exit()` after pip install,

![image](https://user-images.githubusercontent.com/10098314/198062608-f8893fcb-9560-433a-9e03-92a5cd9a8d0b.png)

 but the kernel restatrs automatocally and you can then run,k, e.g. `import spectrochempy as scp`: 

![image](https://user-images.githubusercontent.com/10098314/198063573-34b97c14-9d87-4927-a394-94dfa2acd1a3.png)
Regarding your Mac: can you please show your installed versions ? `conda list`

Thanks
